network/kademlia_load_balancer_test.go

```diff
@@ -0,0 +1,415 @@
+			}
+		}
+	}
+
+}
+
+func newTestKademliaBackend(address string) *testKademliaBackend {
+	return &testKademliaBackend{
+		baseAddr:       pot.NewAddressFromString(address),
+		addedChannel:   gopubsub.New(),
+		removedChannel: gopubsub.New(),
+		bins:           make(map[int][]*Peer),
+	}
+}
+
+// BaseAddr returns the node address of the kademlia table. This base address is the pivot address to which other addresses
+// are sorted with respect to the proximity order function.
+func (tkb testKademliaBackend) BaseAddr() []byte {
+	return tkb.baseAddr
+}
+
+// SubscribeToPeerChanges returns a subscription to changes in the kademlia peers. It contains channels to notify about
+// peers added/removed from this kademlia.
+func (tkb *testKademliaBackend) SubscribeToPeerChanges() (addedSub *gopubsub.Subscription, removedPeerSub *gopubsub.Subscription) {
+	addedSub = tkb.addedChannel.Subscribe()
+	removedPeerSub = tkb.removedChannel.Subscribe()
+	return
+}
+
+// EachBinDescFiltered ignores capKey as in this test context it won't be used. It ignores base as EachBinDesc ignores it.
+func (tkb testKademliaBackend) EachBinDescFiltered(base []byte, capKey string, minProximityOrder int, consumer PeerBinConsumer) error {
+	tkb.EachBinDesc(base, minProximityOrder, consumer)
+	return nil
+}
+
+// EachBinDesc iterates bin in the table in descending po order (from nearest to furthest). Base is always supposed to be
+// node address in this test context. For each bin found, the provided PeerBinConsumer function is called.
+func (tkb testKademliaBackend) EachBinDesc(_ []byte, minProximityOrder int, consumer PeerBinConsumer) {
+	type poPeers struct {
+		po    int
+		peers []*Peer
+	}
+	var poPeersList []poPeers
+	for po, peers := range tkb.bins {
+		poPeersList = append(poPeersList, poPeers{po: po, peers: peers})
+	}
+	sort.Slice(poPeersList, func(i, j int) bool {
+		return poPeersList[i].po > poPeersList[j].po
+	})
+	for _, aPoPeers := range poPeersList {
+		peers := aPoPeers.peers
+		po := aPoPeers.po
+		if peers != nil && po >= minProximityOrder {
+			bin := &PeerBin{
+				ProximityOrder: po,
+				Size:           len(peers),
+				PeerIterator: func(consumePeer PeerConsumer) bool {
+					for _, peer := range peers {
+						if !consumePeer(&entry{conn: peer}) {
+							return false
+						}
+					}
+					return true
+				},
+			}
+			if !consumer(bin) {
+				return
+			}
+		}
+	}
+}
+
+func (tkb *testKademliaBackend) addPeer(peer *Peer) {
+	po, _ := Pof(peer.Address(), tkb.baseAddr, 0)
+	if tkb.bins[po] == nil {
+		if po > tkb.maxPo {
+			tkb.maxPo = po
+		}
+		tkb.bins[po] = make([]*Peer, 0)
+	}
+	tkb.bins[po] = append(tkb.bins[po], peer)
+	tkb.addedChannel.Publish(newPeerSignal{
+		peer: peer,
+		po:   po,
+	})
+	// As the subscribers to add peer are asynchronous, we will sleep here to allow them to execute.
+	// Because this will be used in tests several times, we wait here so the code is not polluted with Sleep calls.
+	time.Sleep(100 * time.Millisecond)
```

Comment:
This sleep is hoping that the scheduler will schedule subscription goroutines in the background. This may not happen, especially on Travis, making tests flaky. Consider a retry mechanism for synchronisation or some more advanced one.